### PR TITLE
Introduce a new configuration that skip storing audit payload if payload size exceed limit and skip storing null fields for audit payload

### PR DESCRIPTION
**Pull request by maytasm**
Introduce a new configuration that skip storing audit payload if payload size exceed limit and skip storing null fields for audit payload

### Description

Audit log such as coordinator.compaction.config can get very large as datasources and their configurations churn. This can cause problems such as 
- Metadata database taking up too much space / running out of storage space
- Configuration change (such as compaction configuration) failing as we are not able to insert new row into audit table due to exceeding the database size limit (max_allowed_packet, etc) 

This PR introduce a new configuration that skip storing audit payload if payload size exceed configured limit. The default value is set to -1 which disables this check, meaning Druid will always store audit payload regardless of it's size. This PR also skip storing null fields for audit payload in the effort of saving storage space as those are unnecessary. 

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Review by kfaraz — COMMENTED** [review]
Perhaps a more readable name for the config would be `druid.audit.manager.maxPayloadSizeBytes`? I don't think the exact behaviour (skipping the store in the case) needs to be a part of the config name.

**Review by mghosh4 — APPROVED** [review]

**Comment by maytasm** [comment]
> Nit: Perhaps a more readable name for the config would be `druid.audit.manager.maxPayloadSizeBytes`? I don't think the exact behaviour (skipping the store in this case) needs to be a part of the config name.

I like that name better. Done

**Comment by jihoonson** [comment]
Added 'Design Review' as it introduces a new user-facing configuration.

File: docs/configuration/index.md
Comment (suneet-s) [inline docs/configuration/index.md:341]: nit: This way we can avoid the update to the `.spelling` file

```suggestion
|`druid.audit.manager.maxPayloadSizeBytes`|The maximum size of audit payload, in bytes, to store in Druid's metadata store audit table. If the size of audit payload exceed this value, the audit log would be stored with a messaging indicating that the payload was omitted instead. Setting `maxPayloadSizeBytes` to -1 (default value) disables this check, meaning Druid will always store audit payload regardless of it's size.|-1|
```

File: core/src/main/java/org/apache/druid/audit/AuditManager.java
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/audit/AuditManager.java:31]: javadocs please

File: core/src/main/java/org/apache/druid/common/config/ConfigSerde.java
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/common/config/ConfigSerde.java:28]: I don't understand this part of the code so well, could you add a javadoc and talk about what the expectations of this function are and how they're different from  `serializeToString` above. Maybe javadocs will help clarify this for me

File: core/src/main/java/org/apache/druid/common/config/JacksonConfigManager.java
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/common/config/JacksonConfigManager.java:55]: I think this object should be bound in the `JacksonModule` as a singleton.

Even though JacksonConfigManager is a singleton, if this binding changes in the future, we could inadvertently create an ObjectMapper per JacksonConfigManager instead of a singleton that can be shared across the process.
Otherwise, each time we create a new JacksonConfigMapper, we will create a new ObjectMapper, even though, jsonMapper is a singleton.

**Review by suneet-s — COMMENTED** [review]
Approach LGTM.

The defaults seem good as well 👍

File: docs/configuration/index.md
Comment (maytasm) [reply docs/configuration/index.md:341]: Done

File: core/src/main/java/org/apache/druid/audit/AuditManager.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/audit/AuditManager.java:31]: Done

File: core/src/main/java/org/apache/druid/common/config/ConfigSerde.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/common/config/ConfigSerde.java:28]: Done

File: core/src/main/java/org/apache/druid/common/config/JacksonConfigManager.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/common/config/JacksonConfigManager.java:55]: Done

File: server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java
Comment (FrankChen021) [inline server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java:35]: Could this property declared as type of  `HumanReadableBytes` ?

**Comment by FrankChen021** [comment]
I have a question. Users have no idea of the size of payload, are there any suggestions that tell users how to set this configuration ? Can we skip storing payload based on value of `type` ?

File: server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java
Comment (maytasm) [reply server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java:35]: `HumanReadableBytes` won't work here. I am using the value `-1` as disabled this check / unlimited max payload size. This is the same as config like maxBytesInMemory where setting maxBytesInMemory to -1 disables the check.

**Comment by maytasm** [comment]
> I have a question. Users have no idea of the size of payload, are there any suggestions that tell users how to set this configuration ? Can we skip storing payload based on value of `type` ?

In another PR, I am adding an auto cleanup based on size limit for the druid_audit metadata table. This work similar to druid.indexer.logs.kill.* which periodically clean up task entries in the druid_tasks table. Basically, user will be able to configured the max number of entries they want to store in the druid_audit metadata table, and Druid will automatically remove from oldest entry (FIFO) when the table entries exceed this size limit. Using both of these configuration, Druid operator will be able to control the size of the audit metadata table. For example, if you want to limit your audit table metadata to 20GB (that is how much storage you have on your machine, etc.), and say you want keep `x` number of last audit entries, you can then set `maxPayloadSizeBytes` to `20GB / x`

File: core/src/main/java/org/apache/druid/guice/annotations/JsonOnlyNonNullValueSerialization.java
Comment (jihoonson) [inline core/src/main/java/org/apache/druid/guice/annotations/JsonOnlyNonNullValueSerialization.java:35]: This name seems quite too long to me :sweat_smile: How about `JsonNonNull`?

File: server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java
Comment (FrankChen021) [reply server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java:35]: > Does HumanReadableBytes not work with -1?

`HumanReadableBytes` does not work with negative value from config since I think there's no such scenario that a size could be negative. But it can accept negative value from code. 

Here I think there're some ways:
1. if -1 is allowed in configuration, we could remove that constraint in `HumanReadableBytes`, and apply `HumanReadableBytesRange` annotation to make sure its value is in valid range
2. or, if we leave that constraint, user must delete this configuration to disable the check
3. or, provide an independent flag to enable or disable the check as suggested by @jihoonson 

IMO, 3rd is better.

> I noticed we have no unit tests for HumanReadableBytes. It would be nice to have some 🙂.

`HumanReadableBytes` has been covered by UT in [HumanReadableBytesTest](https://github.com/apache/druid/blob/master/core/src/test/java/org/apache/druid/java/util/common/HumanReadableBytesTest.java) file.

Comment (maytasm) [reply server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java:35]: @jihoonson  Ah, I was wrong. `HumanReadableBytes` does allow `-1`. For example, CaffeineCacheConfig's sizeInBytes also uses `HumanReadableBytes` and set default to `-1`. Changed this to use `HumanReadableBytes`. However, I kept the default value as `-1` which is use for disabling the check. This is consistent with a few other configs we have in Druid right now. We can discuss on changing all of the configs and standardize disable/enable in a separate PR.

File: core/src/main/java/org/apache/druid/audit/AuditManager.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/audit/AuditManager.java:35]: Done

**Comment by FrankChen021** [comment]
> > I have a question. Users have no idea of the size of payload, are there any suggestions that tell users how to set this configuration ? Can we skip storing payload based on value of `type` ?
> 
> In another PR, I am adding an auto cleanup based on size limit for the druid_audit metadata table. This work similar to druid.indexer.logs.kill.* which periodically clean up task entries in the druid_tasks table. Basically in my other PR, user will be able to configured the max number of entries they want to store in the druid_audit metadata table, and Druid will automatically remove from oldest entry (FIFO) when the table entries exceed this size limit. Using both of these configuration (this PR and the other PR), Druid operator will be able to control the size of the audit metadata table. For example, if you want to limit your audit table metadata to 20GB (i.e. that is how much storage you have on your machine), and say you want keep `x` number of last audit entries, you can then set `maxPayloadSizeBytes` to `20GB / x`
> This make sure that the audit table will not grow uncontrollably and bringing down / effecting your whole Druid cluster.

That makes sense. Do you have any plan to put these explanation in doc by your another PR ?

File: core/src/main/java/org/apache/druid/common/config/ConfigSerde.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/common/config/ConfigSerde.java:33]: Oops. Updated.

**Comment by maytasm** [comment]
> > > I have a question. Users have no idea of the size of payload, are there any suggestions that tell users how to set this configuration ? Can we skip storing payload based on value of `type` ?
> > 
> > 
> > In another PR, I am adding an auto cleanup based on size limit for the druid_audit metadata table. This work similar to druid.indexer.logs.kill.* which periodically clean up task entries in the druid_tasks table. Basically in my other PR, user will be able to configured the max number of entries they want to store in the druid_audit metadata table, and Druid will automatically remove from oldest entry (FIFO) when the table entries exceed this size limit. Using both of these configuration (this PR and the other PR), Druid operator will be able to control the size of the audit metadata table. For example, if you want to limit your audit table metadata to 20GB (i.e. that is how much storage you have on your machine), and say you want keep `x` number of last audit entries, you can then set `maxPayloadSizeBytes` to `20GB / x`
> > This make sure that the audit table will not grow uncontrollably and bringing down / effecting your whole Druid cluster.
> 
> That makes sense. Do you have any plan to put these explanation in doc by your another PR ?

Yep. This is the other PR https://github.com/apache/druid/pull/11084. Right now it only has the doc change and the code implementation is coming soon. Please let me know if you think any docs is missing. I will add a section on how both configs can be use together to control the size of the metadata tables.

File: core/src/main/java/org/apache/druid/guice/annotations/JsonOnlyNonNullValueSerialization.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/guice/annotations/JsonOnlyNonNullValueSerialization.java:29]: Done

File: core/src/main/java/org/apache/druid/guice/annotations/JsonOnlyNonNullValueSerialization.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/guice/annotations/JsonOnlyNonNullValueSerialization.java:35]: Done 👍

File: core/src/main/java/org/apache/druid/common/config/JacksonConfigManager.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/common/config/JacksonConfigManager.java:86]: Done

File: docs/configuration/index.md
Comment (jihoonson) [inline docs/configuration/index.md:342]: `will be exclude` -> `will exclude`?

File: server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java
Comment (jihoonson) [reply server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java]: @FrankChen021 oh you are right. There is a class for unit tests. Sorry, a filter that excludes all test classes might be on in my intellij when I searched last time :sweat:

**Review by jihoonson — COMMENTED** [review]
Thanks for the quick update. I left a couple of trivial comments.

File: core/src/main/java/org/apache/druid/audit/AuditManager.java
Comment (maytasm) [reply core/src/main/java/org/apache/druid/audit/AuditManager.java:36]: Done

File: server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java
Comment (maytasm) [reply server/src/main/java/org/apache/druid/server/audit/SQLAuditManagerConfig.java:36]: Done. Also updated the docs.

File: docs/configuration/index.md
Comment (maytasm) [reply docs/configuration/index.md:342]: Done

File: docs/configuration/index.md
Comment (jihoonson) [inline docs/configuration/index.md:341]:
```suggestion
|`druid.audit.manager.maxPayloadSizeBytes`|The maximum size of audit payload to store in Druid's metadata store audit table. If the size of audit payload exceeds this value, the audit log would be stored with a message indicating that the payload was omitted instead. Setting `maxPayloadSizeBytes` to -1 (default value) disables this check, meaning Druid will always store audit payload regardless of it's size. Setting to any negative number other than `-1` is invalid. Human-readable format is supported, see [here](human-readable-byte.md).  |-1|
```

**Review by jihoonson — COMMENTED** [review]
LGTM. I left a suggestion on typos in docs.

**Comment by maytasm** [comment]
@FrankChen021 I have changed the config to `HumanReadableBytes`. Do you have any other concern or blocker before merging this PR? Thanks!

**Review by suneet-s — APPROVED** [review]
The configs, defaults and approach LGTM

I have not dug into tests, but since this design review, another committer can confirm this.

**Review by jihoonson — APPROVED** [review]
Thanks! +1 after CI.

**Comment by maytasm** [comment]
@jihoonson @suneet-s @FrankChen021 Thanks all for the review.

@FrankChen021 Let me know if you have any followup comments or concerns.

**Pull request merged**
